### ISSUE TITLE
Deadlock when running statements created outside of transaction in transaction

### ISSUE THREAD

**Issue by sightsnow** [simolus3/moor#2873]
**Describe the bug**

code:
```
  Future<void> changeNoteParent(
      {required String noteId,
      required String parentNoteId,
      required String newParentNoteId,
      required int newPositon}) async {
    final branch = await getBranch(noteId, parentNoteId);
    final sql = db.update(db.branches)
      ..where((branches) =>
          branches.noteId.equals(noteId) &
          branches.parentNoteId.equals(parentNoteId));
    final data = BranchesCompanion(
      parentNoteId: Value(newParentNoteId),
      position: Value(newPositon),
      utcModified: Value(TimeTool.utcms),
    );
    await db.transaction(() async {
      await db.customStatement(
        'UPDATE branches SET position = position + 1 WHERE parentNoteId = ? AND position >= ?',
        [newParentNoteId, newPositon],
      );
      await sql.write(data);
      await db.customStatement(
        'UPDATE branches SET position = position - 1 WHERE parentNoteId = ? AND position >= ?',
        [parentNoteId, branch.position],
      );
    });
  }
```

log:
```
flutter: Drift: Sent SELECT * FROM "branches" WHERE "noteId" = ? AND "parentNoteId" = ?; with args [4Rj1ynrk23yKtMkWb1qhQ, EWcj_F9G8Tag_drUt58yO]
flutter: Drift: Sent BEGIN TRANSACTION with args []
flutter: Drift: Sent UPDATE branches SET position = position + 1 WHERE parentNoteId = ? AND position >= ? with args [HJvAmgl-clPpnTdJzZGnp, 1]

```

As shown above, only one SQL statement is performed, and it is not executed down.

**Comment by simolus3** [comment]
The problem appears to be creating the `sql` statement outside of the transaction and then running it inside of the transaction. By moving the definition of `sql` into the transaction, it completes.

This is still a flaw in drift, the statement should run in the transaction if `run` is called in there.

**Comment by sightsnow** [comment]
> The problem appears to be creating the `sql` statement outside of the transaction and then running it inside of the transaction. By moving the definition of `sql` into the transaction, it completes.
> 
> This is still a flaw in drift, the statement should run in the transaction if `run` is called in there.

I tried it. yes. now it can run.

**Comment by simolus3** [comment]
The drift bug has been fixed in 7f7d2ab1b0a03383eb6acd30ca376026720ce347,  so the original snippet will work in the next drift version.
